aws/aws-toolkit-vscode
# Added Stack selection to SAM Deploy Wizard

## Description

The SAM Application Deploy flow used to ask users to type in the CloudFormation stack they would like to deploy to. This has been replaced with a list that allows users to choose from an existing stack. A button allows users to enter a new stack name.

The selector opens in a "loading" state, and the stacks populate into the list as they are retrieved from the account. This typically happens very fast, but supports situations where loading may take a while.

Also in this change:
* The deploy wizard tests related to s3 validation were incorrectly returning test pass results. The tests have been revised to exercise the validation method.
* Reusable CloudFormation Stack selector and input prompter
* Reusable async-iter Loading component that hooks to events, enabling multiple consumers from a single iterator

## Related Issue(s)

#532 

## Testing

* existing tests, new tests
* Run the Toolkit, go through the deploy SAM Application workflow:
  * happy path
  * load a region containing no stacks -- see placeholder entry
  * load a region I don't have access to - see error entry

## Screenshots (if appropriate)

Typical flow - also demonstrates:
* clicking from the selector to entering a new stack name, and back
* Typing in the name of an existing stack is validated against
![cf-stack-select](https://user-images.githubusercontent.com/39839589/60058418-2a95c100-969d-11e9-9f0e-a27c2163a4b8.gif)

Illustrates adding a placeholder entry to the picker, when no stacks were found. Clicking the entry behaves like Cancel/Back/Esc
![cf-stack-empty](https://user-images.githubusercontent.com/39839589/60058440-44370880-969d-11e9-97e7-b91e50530393.gif)

Illustrates an error getting surfaced into the picker. Clicking the entry behaves like Cancel/Back/Esc
![cf-stack-error](https://user-images.githubusercontent.com/39839589/60058455-51ec8e00-969d-11e9-8b71-3a0c463a46b1.gif)


## Checklist





## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
